### PR TITLE
pegasus multi install

### PR DESCRIPTION
**Pull request by rocherroche**
Break dependencies of Pegasus in requirements.txt into the set of minimal dependencies and several sets of optional dependency. This is to allow users avoid installing too many packages that they may not need.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -50,7 +50,12 @@
     extras_require=dict(
         tsne=["fitsne"],
         leiden=["leidenalg"],
-        mkl=["mkl"]
+        louvain=["louvain"],
+        demuxEM=["demuxEM"],
+        scanorama=["scanorama"],
+        gprofiler=["gprofiler-official"],
+        torch=["torch"],
```
Comment (yihming) [inline setup.py]: The `torch` list should also include `harmony-pytorch` and `nmf-torch` packages.

File: pegasus/tools/visualization.py [hunk 2]
```diff
@@ -2,12 +2,15 @@
 import numpy as np
 import scipy
 import umap as umap_module
-import forceatlas2 as fa2
 import uuid
 
 from threadpoolctl import threadpool_limits
 from pegasusio import MultimodalData
 from pynndescent import NNDescent
+try:
```
Comment (yihming) [inline pegasus/tools/visualization.py]: Why do we need Line 10 to 13?

File: pegasus/tools/visualization.py [hunk 3]
```diff
@@ -54,7 +57,11 @@ def calc_tsne(
     if fftw3_loc is None:
         raise Exception("Please install 'fftw3' first to use the FIt-SNE feature!")
 
-    from fitsne import FItSNE
+    try:
+        from fitsne import FItSNE
+    except PackageNotFoundError:
+        print("FItSNE package is not installed")
```
Comment (yihming) [inline pegasus/tools/visualization.py]: Add installation suggestion `pip install fitsne`.

File: pegasus/tools/visualization.py [hunk 4]
```diff
@@ -54,7 +57,11 @@ def calc_tsne(
...
+        exit()
```
Comment (yihming) [inline pegasus/tools/visualization.py:60]: Not needed because this is within the `except` block.

File: pegasus/tools/visualization.py [hunk 5]
```diff
@@ -136,18 +143,22 @@ def calc_force_directed_layout(
     TODO: Typing
     """
     G = construct_graph(W)
+    try:
+        import forceatlas2 as fa2
+    except PackageNotFoundError:
+        print("Please install module forceatlas2")
```
Comment (yihming) [inline pegasus/tools/visualization.py]: 1. Correct the package name to `forceatlas2-python` in your `print` statement, although when importing it in Python, its module name is `forceatlas2`.
2. Add installation suggestion `pip install forceatlas2-python`.

File: pegasus/tools/visualization.py [hunk 6]
```diff
@@ -136,18 +143,22 @@ def calc_force_directed_layout(
...
+        exit()
```
Comment (yihming) [inline pegasus/tools/visualization.py:146]: Not needed because this is within the `except` block.

File: setup.py [hunk 7]
```diff
@@ -50,7 +50,13 @@
...
         tsne=["fitsne"],
         leiden=["leidenalg"],
-        mkl=["mkl"]
+        louvain=["louvain"],
+        scanorama=["scanorama"],
+        gprofiler=["gprofiler-official"],
+        torch=["torch"],
+        forceatlas=["forceatlas2-python"],
+        mkl=["mkl"],
+        all=["all"]
```
Comment (yihming) [inline setup.py]: What is `"all"` here?
The purpose of adding this `all` dependency list is to allow user use a one-line command to install all the dependencies (both required and optional) as before, i.e.
```
pip install pegasuspy[all]
```
will install everything.

File: pegasus/tools/visualization.py (hunk 2 above)
Comment (yihming) [reply pegasus/tools/visualization.py]: I understand your purpose now: you want to raise `PackageNotFoundError` which is not in Python's default exceptions but from `importlib_metadata`.

Is there any particular reason you want to use `PackageNotFoundError`, while either `ImportError` or its single subclass `ModuleNotFoundError` from Python's default exceptions could cover the case? I mean, if you use `ImportError` or `ModuleNotFoundError`, you don't need to write any `import` statement.

File: setup.py [hunk 8]
```diff
@@ -50,7 +50,13 @@
     extras_require=dict(
         tsne=["fitsne"],
         leiden=["leidenalg"],
-        mkl=["mkl"]
+        louvain=["louvain"],
+        scanorama=["scanorama"],
+        gprofiler=["gprofiler-official"],
```
Comment (yihming) [inline setup.py]: I don't think `gprofiler` is in your task card. And we have already removed it from Pegasus.

**Review by yihming — CHANGES_REQUESTED** [review]
Please resolve the issues I wrote inline.

Besides, to make your PR pass the CI test, when you are done with `all` list in `extras_require`, just change Line 31 in `.github/workflows/ci-test.yml` to
```
python -m pip install -e .[all]
```

File: pegasus/tools/visualization.py (hunk 4 above)
Comment (rocherroche) [reply pegasus/tools/visualization.py:60]: yes but it we catch the exception without doing anything, the script would continue its execution..

File: pegasus/tools/visualization.py (hunk 6 above)
Comment (rocherroche) [reply pegasus/tools/visualization.py:146]: if we catch the exception without doing anything than printing, the script would continue its execution..

File: setup.py [hunk 9]
```diff
@@ -50,7 +50,12 @@
     extras_require=dict(
         tsne=["fitsne"],
         leiden=["leidenalg"],
```
Comment (yihming) [inline setup.py]: Due to the Trello card assigned to you, `leidenalg` is in the required dependencies, and thus this `leiden` list should removed from `extras_require`.

File: setup.py [hunk 10]
```diff
@@ -50,7 +50,12 @@
...
-        mkl=["mkl"]
+        louvain=["louvain"],
+        scanorama=["scanorama"],
+        torch=["torch", "harmony-pytorch", "nmf-torch"],
+        forceatlas=["forceatlas2-python"],
+        mkl=["mkl"],
+        all=["fitsne", "leidenalg", "louvain", "scanorama", "torch", "forceatlas2-python", "mkl"]
```
Comment (yihming) [inline setup.py]: Remove `leidenalg` because it should be in the required dependencies.

Comment (yihming) [reply setup.py]: Please remove `mkl` from `all` list, too.

File: pegasus/tools/visualization.py (hunk 4 above)
Comment (yihming) [reply pegasus/tools/visualization.py:60]: Ah, I remember that you have mentioned it before. Thank you for the reminder!

To make the print statement more bold, you may want to add "Exception" ahead, i.e.
```
print("Exception: FItSNE package is not installed, please install fitsne: pip install fitsne")
```

File: pegasus/tools/visualization.py (hunk 6 above)
Comment (yihming) [reply pegasus/tools/visualization.py:146]: Similarly as above, add "Exception:" at the beginning of the print statement in the except block.

**Review by yihming — CHANGES_REQUESTED** [review]

**Review by yihming — APPROVED** [review]

**Pull request merged**